### DISCUSSION TITLE
How to Replace a Foreach Loop with LanguageExt in Async Code Involving Database Operations?

### DISCUSSION THREAD

**Discussion by padnom** [louthy/language-ext#1404] [Q&A] [answered]
Hello,

First of all, I want to say that LanguageExt is a fantastic library

I'm trying to adapt my code to use LanguageExt's functional programming constructs, but I'm not sure how to replace my foreach loop with a more functional approach using LanguageExt. My code involves asynchronous operations that may add data to a database. Here's my current code:

```cs
foreach (var member in group.GetMembers(true))
{
    await _packLicenseRepository.GetCollaborateurAsync(member.Guid.Value, cancellationToken)
        .BindAsync(collaborateur => Person.CreatePerson(collaborateur.CollaborateurPk))
        .BindAsync(person => serviceStore.AddPerson(pack.PackName, person));
}

```
I would like to refactor this code to eliminate the foreach loop and use LanguageExt's functional methods instead, considering that I'm adding entries to a database.

Additionally, I want the processing to continue even if some operations fail, and I wish to collect a list of any errors that occur during the process. Could someone please guide me on how to achieve this?

Thank you!

**Comment by louthy** [comment]
You will need to do a bit of reading on [Traversables](https://paullouth.com/higher-kinds-in-csharp-part6-traversables/); but also you should understand applicatives, foldables, and functors.  Traversables can work with foldable-sequences of effects and leverage their applicative behaviours to run operations concurrently, but also collect multiple errors.

You're not going to get that functionality from the `Task` extension because these don't support the new traits and therefore don't support `Traverse`, so you'll need to use the fully traversable types like `Iterable<A>` and `Seq<A>`; and applicative types like `Eff<A>` (think of `Eff<A>` as a more advanced `Task<A>`).

**Comment by padnom** [comment]
Thx for your response, There's a lot to read also for the Eff part.
If I read your documentation correctly regarding traversables, it's preferable that I use map to be able to retrieve both errors and successes.

Do you have a code example with async calls to a database?

**Reply by louthy** [reply, marked as answer]
First off, make sure you have the latest beta, it needed some changes to support this for `IO`-based operations.

No, don't use `Map`, use `Traverse`:
```c#
// Two failing Eff effects, note how the Task is lifted into the Eff.  
var eff1 = liftEff(e => Task.FromException<int>(Error.New("failed 1")));
var eff2 = liftEff(e => Task.FromException<int>(Error.New("failed 2")));

// Put the two effects into a Seq and Traverse
var eff3 = Seq(eff1, eff2).Traverse(x => x);

// Run the effect to get the result
var res  = await eff3.RunAsync();

Console.WriteLine(res);   // Fail([failed 1, failed 2])
```
I note you've changed the requirements to "retrieve both errors and successes."; the above will only return either the successes or the failures, not both.

You can use `Partition` if you need to collect both:
```c#
var effs = Seq(eff1, eff2);
var res = effs.Map(eff => eff.Run()).Partition();  // (Seq<Error> Fails, Seq<int> Succs)
```
It partitions the results into a tuple of `Fails` and `Succs`.

**Reply by padnom** [reply]
Thx a lot for you time I wil try to switch to the V5 !
